crates/turborepo-lib/src/run/scope/filter.rs

```diff
@@ -50,4 +64,1082 @@ impl PackageInference {
+    ///
+    /// Example:
+    /// a -> b -> c
+    /// a -> d
+    ///
+    /// filter(b) = {a, b, c}
+    /// filter(d) = {a, d}
+    fn filter_subtrees_with_selector(
+        &self,
+        selector: &TargetSelector,
+    ) -> Result<HashSet<String>, ResolutionError> {
+        let mut entry_packages = HashSet::new();
+
+        for (name, info) in self.pkg_graph.workspaces() {
+            if selector.parent_dir == AnchoredSystemPathBuf::default() {
+                entry_packages.insert(name.to_string());
+            } else {
+                let path = self
+                    .turbo_root
+                    .join_unix_path(selector.parent_dir.to_unix().unwrap())
+                    .unwrap();
+                let matcher = wax::Glob::new(path.as_str())?;
+                let p = info.package_json_path.restore_anchor(self.turbo_root);
+                let matches = matcher.is_match(p.as_std_path());
+
+                if matches {
+                    entry_packages.insert(name.to_string());
+                }
+            }
+        }
+
+        // if we have a filter, use it to filter the entry packages
+        let filtered_entry_packages = if !selector.name_pattern.is_empty() {
+            match_package_names(&selector.name_pattern, entry_packages)?
+        } else {
+            entry_packages
+        };
+
+        let mut roots = HashSet::new();
+        let mut matched = HashSet::new();
+        let changed_packages =
+            self.packages_changed_in_range(&selector.from_ref, selector.to_ref())?;
+
+        for package in filtered_entry_packages {
+            if matched.contains(&package) {
+                roots.insert(package);
+                continue;
+            }
+
+            let workspace_node = package_graph::WorkspaceNode::Workspace(
+                package_graph::WorkspaceName::Other(package.clone()),
+            );
+
+            let dependencies = self.pkg_graph.dependencies(&workspace_node);
+
+            for changed_package in &changed_packages {
+                if !selector.exclude_self && package.eq(changed_package) {
+                    roots.insert(package);
+                    break;
+                }
+
+                let changed_node = package_graph::WorkspaceNode::Workspace(
+                    package_graph::WorkspaceName::Other(changed_package.clone()),
+                );
+
+                if dependencies.contains(&changed_node) {
+                    roots.insert(package.clone());
+                    matched.insert(package);
+                    break;
+                }
+            }
+        }
+
+        Ok(roots)
+    }
+
+    fn filter_nodes_with_selector(
+        &self,
+        selector: &TargetSelector,
+    ) -> Result<HashSet<String>, ResolutionError> {
+        let mut entry_packages = HashSet::new();
+        let mut selector_used = false;
+
+        let path = self
+            .turbo_root
+            .join_unix_path(selector.parent_dir.to_unix().unwrap())
```

Comment:
Here, and elsewhere, can you either remove the `.unwrap()`, or change it to an `.expect()` with documentation as to why we don't expect to hit it?

Reply:
I have a handful of unwraps I need to fix. I will get round to those asap.